From c8b37e92b4efebd5492e5e1ad4a1b84a641810f8 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Date: Fri, 8 Feb 2013 19:06:49 +0100
Subject: [PORT FROM MAIN]wifi: Support for changing driver mode without FW reloading

BZ: 83709

This is based on Jonathan De Cesco's patch 82732: in order to support SoftAP
and P2P mode with the new driver loading policy (insmod at boot), we need to be
able to change the driver mode without rmmoding the driver.

Change-Id: I9a5fbec5983de3faf3db805405b5254abae07ab6
Orig-Change-Id: If1b808bd60f65d0a9a3b4530a2b11e1c7a20d8bf
Category: device enablement
Domain: CWS.WIFI-Bluewoods
Origin: upstream
Upstream-Candidate: yes
Signed-off-by: Quentin Casasnovas <quentinx.casasnovas@intel.com>
---
 SoftapController.cpp |   67 ++++++++++++++++++++++++++++++++++++++------------
 SoftapController.h   |    2 +
 2 files changed, 53 insertions(+), 16 deletions(-)

diff --git a/SoftapController.cpp b/SoftapController.cpp
index a482784..47904d6 100644
--- a/SoftapController.cpp
+++ b/SoftapController.cpp
@@ -51,6 +51,42 @@ SoftapController::SoftapController()
 SoftapController::~SoftapController() {
 }
 
+int SoftapController::startDriver(char *iface) {
+    int ret;
+
+    ALOGD("Softap driver start");
+
+    if (!iface || (iface[0] == '\0')) {
+        ALOGE("Softap driver start - wrong interface");
+        return -EINVAL;
+    }
+
+    ifc_init();
+    ret = ifc_up(iface);
+    ifc_close();
+
+    usleep(AP_DRIVER_START_DELAY);
+
+    return ret;
+}
+
+int SoftapController::stopDriver(char *iface) {
+    int ret;
+
+    ALOGD("Softap driver stop");
+
+    if (!iface || (iface[0] == '\0')) {
+        ALOGE("Softap driver stop - wrong interface");
+	return -EINVAL;
+    }
+
+    ifc_init();
+    ret = ifc_down(iface);
+    ifc_close();
+
+    return ret;
+}
+
 int SoftapController::startSoftap() {
     pid_t pid = 1;
 
@@ -207,28 +243,27 @@ int SoftapController::fwReloadSoftap(int argc, char *argv[])
 #else
     int i = 0;
     char *fwpath = NULL;
+    char *iface;
 
     if (argc < 4) {
         ALOGE("SoftAP fwreload is missing arguments. Please use: softap <wlan iface> <AP|P2P|STA>");
         return ResponseCode::CommandSyntaxError;
     }
 
-    if (strcmp(argv[3], "AP") == 0) {
-        fwpath = (char *)wifi_get_fw_path(WIFI_GET_FW_PATH_AP);
-    } else if (strcmp(argv[3], "P2P") == 0) {
-        fwpath = (char *)wifi_get_fw_path(WIFI_GET_FW_PATH_P2P);
-    } else if (strcmp(argv[3], "STA") == 0) {
-        fwpath = (char *)wifi_get_fw_path(WIFI_GET_FW_PATH_STA);
-    }
-    if (!fwpath)
-        return ResponseCode::CommandParameterError;
-    if (wifi_change_fw_path((const char *)fwpath)) {
-        ALOGE("Softap fwReload failed");
-        return ResponseCode::OperationFailed;
-    }
-    else {
-        ALOGD("Softap fwReload - Ok");
-    }
+    iface = argv[2];
+    stopDriver(iface);
+
+    wifi_switch_driver_mode(WIFI_AP_MODE);
+
+    /**
+     * Sleep to workaround issue in the brcm driver which is tracked by BZ
+     * 85864. To be removed as soon as BZ85864 is fixed.
+     */
+    sleep(2);
+    startDriver(iface);
+
+    ALOGD("Softap fwReload - done");
+
     return ResponseCode::SoftapStatusResult;
 #endif
 }
diff --git a/SoftapController.h b/SoftapController.h
index 7063067..282b916 100644
--- a/SoftapController.h
+++ b/SoftapController.h
@@ -32,6 +32,8 @@ public:
     SoftapController();
     virtual ~SoftapController();
 
+    int startDriver(char *iface);
+    int stopDriver(char *iface);
     int startSoftap();
     int stopSoftap();
     bool isSoftapStarted();
-- 
1.7.4.1

